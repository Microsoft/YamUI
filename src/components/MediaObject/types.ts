/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum MediaObjectSize {
  XLARGE = 'xLarge',
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
