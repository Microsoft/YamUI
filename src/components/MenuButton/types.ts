/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { ContextualMenuItemType } from 'office-ui-fabric-react/lib/ContextualMenu';

export enum MenuItemType {
  Normal = ContextualMenuItemType.Normal,
  Header = ContextualMenuItemType.Header,
  Divider = ContextualMenuItemType.Divider,
}
