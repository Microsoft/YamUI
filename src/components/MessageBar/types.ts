/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum MessageBarType {
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error',
}
