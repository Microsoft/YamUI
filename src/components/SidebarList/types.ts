/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseIllustration } from '../Illustration';

export type SidebarListSize = 'small' | 'medium' | 'large';

export interface BaseSidebarListItem {
  /**
   * A unique key for the item.
   */
  key: string;

  /**
   * The visible title text.
   */
  title: string;

  /**
   * Metadata text displayed under the title, not displayed at small size.
   */
  description?: string;
}
export interface SidebarListIllustrationItem extends BaseSidebarListItem {
  /**
   * A YamUI Illustration component class. You are responsible for importing and providing the correct size.
   */
  illustration: typeof BaseIllustration;
}
export interface SidebarListImageItem extends BaseSidebarListItem {
  /**
   * Custom image JSX. It should fit the list item's pixel width at the given size, or fill 100% of its container width.
   */
  imageContent: React.ReactNode;
}
export type SidebarListItemWithGraphic = SidebarListIllustrationItem | SidebarListImageItem;

export type SidebarListClickableItem = SidebarListItemWithGraphic & {
  /**
   * Callback to be triggered when the item is clicked/invoked.
   */
  onClick(key: BaseSidebarListItem['key']): void;
};
export type SidebarListLinkItem = SidebarListItemWithGraphic & {
  /**
   * The URL this item links to.
   */
  href: string;

  /**
   * If true, opens the link securely in a new window.
   */
  newWindow?: boolean;
};
export type SidebarListItem = SidebarListClickableItem | SidebarListLinkItem;
