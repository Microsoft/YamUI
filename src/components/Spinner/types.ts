/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum SpinnerColor {
  /**
   * Light theme.
   */
  LIGHT = 'light',
  /**
   * Light theme spinner color with gray text.
   */
  METADATA = 'metadata',
  /**
   * Dark theme.
   */
  DARK = 'dark',
}
export enum SpinnerSize {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
