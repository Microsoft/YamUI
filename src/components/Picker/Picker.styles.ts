/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets, ITheme } from 'office-ui-fabric-react/lib/Styling';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';
export interface PickerClassNameProps {
  theme: ITheme;
}

export const getClassNames = memoizeFunction((classNameProps: PickerClassNameProps) => {
  const { theme } = classNameProps;
  return mergeStyleSets({
    selectedItem: {
      backgroundColor: theme.semanticColors.menuItemBackgroundHovered,
      borderRadius: '2px',
      border: `solid 1px ${theme.semanticColors.bodyDivider}`,
      margin: '2px',
      padding: '2px 4px',
      height: '26px',
      color: theme.semanticColors.bodyText,
      cursor: 'default',
    },
    selectedItemCloseButton: {
      cursor: 'pointer',
      display: 'inline-block',
    },
    suggestionItem: {
      width: '100%',
    },
  });
});
