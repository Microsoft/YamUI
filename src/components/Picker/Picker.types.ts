/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { IBasePickerProps } from 'office-ui-fabric-react/lib/Pickers';
import { SuggestionItem } from '../SuggestionsList/SuggestionsListItemContent';

export { IBasePickerProps };

export interface BasePickerProps<T> extends BaseComponentProps {
  /**
   * This is where you provide the items to be rendered in the autocomplete dropdown.
   * The current filter text and list of already-selected items will be provided.
   */
  onResolveSuggestions: IBasePickerProps<T>['onResolveSuggestions'];

  /**
   * Called when the list of selected items changes.
   */
  onChange: IBasePickerProps<T>['onChange'];

  /**
   * The list of currently-selected items.
   */
  selectedItems: IBasePickerProps<T>['selectedItems'];

  /**
   * Aria label for the focusable text input.
   */
  inputAriaLabel: string;

  /**
   * Aria label for the close button on selected items.
   */
  removeItemAriaLabel: string;

  /**
   * Visible header text in the autocomplete dropdown.
   */
  suggestionsHeaderText: string;

  /**
   * Message to display when no results are found.
   */
  noResultsFoundText: string;

  /**
   * Disables the picker UI.
   */
  disabled?: boolean;

  /**
   * If provided, limits the number of items which can be selected.
   */
  itemLimit?: number;

  /**
   * Callback event when the UI element is focused.
   */
  onFocus?: IBasePickerProps<T>['onFocus'];

  /**
   * Callback event when the UI element loses focus.
   */
  onBlur?: IBasePickerProps<T>['onBlur'];
}

export type PickerProps = BasePickerProps<SuggestionItem>;
