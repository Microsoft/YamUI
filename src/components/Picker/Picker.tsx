/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { BasePicker, IBasePickerProps, IPickerItemProps } from 'office-ui-fabric-react/lib/Pickers';
import Block, { TextSize } from '../Block';
import { FixedGridRow, FixedGridColumn } from '../FixedGrid';
import SuggestionItemContent, { SuggestionItem } from '../SuggestionsList/SuggestionsListItemContent';
import RemoveIcon from '../Icon/icons/Cancel';
import { getClassNames } from './Picker.styles';
import './Picker.css'; // Required for the Callout content as there aren't JS styles to override

class SuggestionListBasePicker extends BasePicker<SuggestionItem, IBasePickerProps<SuggestionItem>> {}

export interface BasePickerProps<T> extends BaseComponentProps {
  /**
   * This is where you provide the items to be rendered in the autocomplete dropdown.
   * The current filter text and list of already-selected items will be provided.
   */
  onResolveSuggestions: IBasePickerProps<T>['onResolveSuggestions'];

  /**
   * Called when the list of selected items changes.
   */
  onChange: IBasePickerProps<T>['onChange'];

  /**
   * The list of currently-selected items.
   */
  selectedItems: IBasePickerProps<T>['selectedItems'];

  /**
   * Aria label for the focusable text input.
   */
  inputAriaLabel: string;

  /**
   * Aria label for the close button on selected items.
   */
  removeItemAriaLabel: string;

  /**
   * Visible header text in the autocomplete dropdown.
   */
  suggestionsHeaderText: string;

  /**
   * Message to display when no results are found.
   */
  noResultsFoundText: string;

  /**
   * Disables the picker UI.
   */
  disabled?: boolean;

  /**
   * If provided, limits the number of items which can be selected.
   */
  itemLimit?: number;

  /**
   * Callback event when the UI element is focused.
   */
  onFocus?: IBasePickerProps<T>['onFocus'];

  /**
   * Callback event when the UI element loses focus.
   */
  onBlur?: IBasePickerProps<T>['onBlur'];
}

export type PickerProps = BasePickerProps<SuggestionItem>;

/**
 * Pickers are used to pick one or more entities (e.g. people, groups, topics) from a list with typeahead capabilities.
 */
export default class Picker extends React.Component<PickerProps> {
  public render() {
    const {
      className,
      disabled,
      itemLimit,
      onFocus,
      onBlur,
      onResolveSuggestions,
      onChange,
      selectedItems,
      inputAriaLabel,
      suggestionsHeaderText,
      noResultsFoundText,
    } = this.props;

    return (
      <SuggestionListBasePicker
        className={join(['y-picker', className])}
        onResolveSuggestions={onResolveSuggestions}
        pickerSuggestionsProps={{
          suggestionsHeaderText,
          noResultsFoundText,
        }}
        itemLimit={itemLimit}
        disabled={disabled}
        inputProps={{
          onBlur,
          onFocus,
          'aria-label': inputAriaLabel,
        }}
        onRenderSuggestionsItem={this.getSuggestionItemContent}
        onRenderItem={this.getSelectedItem}
        onChange={onChange}
        selectedItems={selectedItems}
      />
    );
  }

  private getSuggestionItemContent = (item: SuggestionItem) => {
    const classNames = getClassNames();

    return <SuggestionItemContent item={item} className={classNames.suggestionItem} />;
  };

  private getSelectedItem = (props: IPickerItemProps<SuggestionItem>) => {
    const classNames = getClassNames();

    return (
      <div
        role={'listitem'}
        className={classNames.selectedItem}
        key={props.item.id}
        data-selection-index={props.index}
        data-is-focusable={!props.disabled}
      >
        <Block textSize={TextSize.MEDIUM_SUB}>
          <FixedGridRow>
            <FixedGridColumn fixed={true}>
              <span>{props.item.name}</span>
            </FixedGridColumn>
            <FixedGridColumn fixed={true}>
              {!props.disabled && (
                <div
                  className={classNames.selectedItemCloseButton}
                  onClick={props.onRemoveItem}
                  aria-label={this.props.removeItemAriaLabel}
                >
                  <RemoveIcon block={true} />
                </div>
              )}
            </FixedGridColumn>
          </FixedGridRow>
        </Block>
      </div>
    );
  };
}
