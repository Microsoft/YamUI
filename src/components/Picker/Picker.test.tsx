/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { SuggestionItem, SuggestionsListItemTemplate } from '../SuggestionsList/types';
import Picker, { PickerProps } from './index';
import { IPickerItemProps } from 'office-ui-fabric-react/lib/Pickers';

describe('<Picker />', () => {
  let component: ShallowWrapper<PickerProps>;
  let onChange: jest.Mock<Function>;

  const suggestions: SuggestionItem[] = [
    {
      template: SuggestionsListItemTemplate.AVATAR_ROUND,
      id: 1,
      name: 'Cristiana Cline',
      description: 'Data Analyst',
    },
    {
      template: SuggestionsListItemTemplate.AVATAR_ROUND,
      id: 2,
      name: 'Cristian Sayers',
      description: 'Front-End Developer',
      imageUrl: 'user.png',
    },
  ];

  const getPicker = (overrides?: Partial<PickerProps>) => {
    onChange = jest.fn();
    const props = {
      onChange,
      onResolveSuggestions: () => suggestions,
      selectedItems: [],
      inputAriaLabel: 'People to mention',
      removeItemAriaLabel: 'Remove this item',
      suggestionsHeaderText: 'People',
      noResultsFoundText: 'No results found',
      ...overrides,
    };
    return <Picker {...props} />;
  };

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(getPicker());
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(getPicker({ className: 'TEST_CLASSNAME' }));
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when rendering a suggestion item', () => {
    beforeEach(() => {
      component = shallow(getPicker());
    });

    it('matches its snapshot', () => {
      const onRenderSuggestionsItem = component
        .find('SuggestionListBasePicker')
        .prop<(item: SuggestionItem) => React.ReactNode>('onRenderSuggestionsItem');
      const someItem = suggestions[0];
      expect(onRenderSuggestionsItem(someItem)).toMatchSnapshot();
    });
  });

  describe('when rendering a selected item', () => {
    beforeEach(() => {
      component = shallow(getPicker());
    });

    it('matches its snapshot', () => {
      const onRenderItem = component
        .find('SuggestionListBasePicker')
        .prop<(props: IPickerItemProps<SuggestionItem>) => React.ReactNode>('onRenderItem');
      const someItem: IPickerItemProps<SuggestionItem> = { index: 0, item: suggestions[0] };
      expect(onRenderItem(someItem)).toMatchSnapshot();
    });
  });

  describe('when a suggestion is selected', () => {
    beforeEach(() => {
      component = shallow(getPicker());
      component.find('SuggestionListBasePicker').simulate('change');
    });

    it('matches its snapshot', () => {
      expect(onChange).toHaveBeenCalled();
    });
  });
});
