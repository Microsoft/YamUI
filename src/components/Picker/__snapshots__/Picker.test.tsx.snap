// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Picker /> when rendering a selected item matches its snapshot 1`] = `
<div
  className=

      {
        background-color: #f3f5f8;
        border-radius: 2px;
        border: solid 1px #dde0e6;
        color: #343A41;
        cursor: default;
        height: 26px;
        margin-bottom: 2px;
        margin-left: 2px;
        margin-right: 2px;
        margin-top: 2px;
        padding-bottom: 2px;
        padding-left: 4px;
        padding-right: 4px;
        padding-top: 2px;
      }
  data-is-focusable={true}
  data-selection-index={0}
  role="listitem"
>
  <Block
    textSize="mediumSub"
  >
    <FixedGridRow>
      <FixedGridColumn
        fixed={true}
      >
        <span>
          Cristiana Cline
        </span>
      </FixedGridColumn>
      <FixedGridColumn
        fixed={true}
      >
        <div
          aria-label="Remove this item"
          className=

              {
                cursor: pointer;
                display: inline-block;
              }
        >
          <Cancel
            block={true}
          />
        </div>
      </FixedGridColumn>
    </FixedGridRow>
  </Block>
</div>
`;

exports[`<Picker /> when rendering a suggestion item matches its snapshot 1`] = `
<SuggestionsListItemContent
  className=

      {
        width: 100%;
      }
  item={
    Object {
      "description": "Data Analyst",
      "id": 1,
      "name": "Cristiana Cline",
      "template": "mediaRound",
    }
  }
/>
`;

exports[`<Picker /> with additional className matches its snapshot 1`] = `
<SuggestionListBasePicker
  className="y-picker TEST_CLASSNAME"
  inputProps={
    Object {
      "aria-label": "People to mention",
      "onBlur": undefined,
      "onFocus": undefined,
    }
  }
  onChange={[MockFunction]}
  onRenderItem={[Function]}
  onRenderSuggestionsItem={[Function]}
  onResolveSuggestions={[Function]}
  pickerSuggestionsProps={
    Object {
      "noResultsFoundText": "No results found",
      "suggestionsHeaderText": "People",
    }
  }
  selectedItems={Array []}
/>
`;

exports[`<Picker /> with default options matches its snapshot 1`] = `
<SuggestionListBasePicker
  className="y-picker"
  inputProps={
    Object {
      "aria-label": "People to mention",
      "onBlur": undefined,
      "onFocus": undefined,
    }
  }
  onChange={[MockFunction]}
  onRenderItem={[Function]}
  onRenderSuggestionsItem={[Function]}
  onResolveSuggestions={[Function]}
  pickerSuggestionsProps={
    Object {
      "noResultsFoundText": "No results found",
      "suggestionsHeaderText": "People",
    }
  }
  selectedItems={Array []}
/>
`;
